src/chainlink/VRFCoordinator.ts

```diff
@@ -0,0 +1,115 @@
+}
+
+async function fetchTokenAndCoordinatorAddresses() {
+    const tokenAndCoordinatorAddressesResponse = await fetch(
+        'https://raw.githubusercontent.com/0xpolyflow/polyflow-sdk/master/resources/chainlink_contract_and_coordinator.json'
+    );
+    const tokenAndCoordinatorAddressesJson = await tokenAndCoordinatorAddressesResponse.json();
+    tokenAndCoordinatorAddresses = new Map(Object.entries(tokenAndCoordinatorAddressesJson));
+}
+
+
+export class VRFCoordinator {
+
+    public subscriptions: VRFSubscription[] = [];
+    public address: string = "";
+    public chainlinkTokenContractAddress: string = "";
+    public web3provider: any;
+    public projectChainId: number = 0;
+
+    constructor() { }
+
+    public async init() {
+        this.projectChainId = (await MainApi.instance().fetchProject()).chain_id;
+        await fetchTokenAndCoordinatorAddresses();
+        this.address = tokenAndCoordinatorAddresses.get(this.projectChainId).vrf_coordinator_contract;
+        this.chainlinkTokenContractAddress = tokenAndCoordinatorAddresses.get(this.projectChainId).link_token_contract;
+    }
+
+    public async createSubscription(): Promise<VRFSubscription> {
+        if (!this.web3provider) {
+            await fetchChainlist();
+            this.web3provider = new ethers.providers.JsonRpcProvider(chainlist.get(this.projectChainId));
+        };
+        const subscription = await writeContract(
+            this.address,
+            "createSubscription",
+            [],
+            "0"
+        );
+        console.log(subscription);
+        const action = new ContractCallAction(subscription);
+        const result = await action.present();
+        const transactionHash = result.transactionHash as string;
+        const transactionReceipt = await this.web3provider.getTransactionReceipt(transactionHash);
+        const subId = parseInt(transactionReceipt.logs[0].topics[1]);
+        return await this.getSubscription(subId.toString());
+    }
+
+    public async getSubscription(subId: string): Promise<VRFSubscription> {
+        let subscription = this.subscriptions.find(sub => sub.id === subId);
+        if (subscription) return subscription;
+        subscription = new VRFSubscription();
+        subscription.id = subId;
+        const callRequest = await readContract(
+            this.address,
+            "getSubscription",
+            [{ type: "uint64", value: subId }],
+            ["uint96", "uint64", "address", "address[]"],
+            "0x0"
+        );
+        subscription.balance = callRequest.return_values[0];
+        subscription.requestCount = callRequest.return_values[1];
+        subscription.owner = callRequest.return_values[2];
+        subscription.consumers = Array.from(callRequest.return_values[3]);
+        subscription.coordinatorAddress = this.address;
+        subscription.chainlinkTokenContractAddress = this.chainlinkTokenContractAddress;
+        this.subscriptions.push(subscription);
```

Comment:
It's cool that you're caching subscriptions per subscription ID and then reusing them when getSubscription is called for the same subscription ID. But I'm afraid that if the subscription data changes, we will still hold the old data using this method. For example, if the subscription balance is reduced after we fetch the object, we will still show the old data - as if the balance was greater than it actually is.

Why don't we remove the caching and the subscriptions array and simply fetch the subscription object from the chain every time the function is called?

Additionally, add the `getInfo()` function on the VRFSubscription object to re-read the values again if we want to see if something changed. This function will call the `readContract("getSubscription")` on the VRFCoordinator again and update its state with the latest subscription data, while also returning the same data to the caller.